streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java

```diff
@@ -0,0 +1,373 @@
+            tasksAndActions.add(new TaskAndAction(task, Action.ADD));
+            tasksAndActionsCondition.signalAll();
+        } finally {
+            tasksAndActionsLock.unlock();
+        }
+    }
+
+    private void verifyStateFor(final Task task) {
+        if (task.isActive() && task.state() != State.RESTORING) {
+            throw new IllegalStateException("Active task " + task.id() + " is not in state RESTORING. " + BUG_ERROR_MESSAGE);
+        }
+    }
+
+    @Override
+    public void remove(final Task task) {
+    }
+
+    @Override
+    public Set<StreamTask> getRestoredActiveTasks(final Duration timeout) {
+        final long timeoutMs = timeout.toMillis();
+        final long startTime = time.milliseconds();
+        final long deadline = startTime + timeoutMs;
+        long now = startTime;
+        final Set<StreamTask> result = new HashSet<>();
+        try {
+            while (now <= deadline && result.isEmpty()) {
+                restoredActiveTasksLock.lock();
+                try {
+                    while (restoredActiveTasks.isEmpty() && now <= deadline) {
+                        final boolean elapsed = restoredActiveTasksCondition.await(deadline - now, TimeUnit.MILLISECONDS);
+                        now = time.milliseconds();
+                    }
+                    while (!restoredActiveTasks.isEmpty()) {
+                        result.add(restoredActiveTasks.poll());
+                    }
+                } finally {
+                    restoredActiveTasksLock.unlock();
+                }
+                now = time.milliseconds();
+            }
+            return result;
+        } catch (final InterruptedException e) {
+            // ignore
+        }
+        return result;
+    }
+
+    @Override
+    public List<ExceptionAndTasks> getFailedTasksAndExceptions() {
+        final List<ExceptionAndTasks> result = new ArrayList<>();
+        failedTasks.drainTo(result);
+        return result;
+    }
+
+    @Override
+    public Set<Task> getAllTasks() {
+        tasksAndActionsLock.lock();
+        restoredActiveTasksLock.lock();
+        try {
+            final Set<Task> allTasks = new HashSet<>();
+            allTasks.addAll(tasksAndActions.stream()
+                .filter(t -> t.action == Action.ADD)
+                .map(t -> t.task)
+                .collect(Collectors.toList())
+            );
+            allTasks.addAll(stateUpdaterThread.getAllUpdatingTasks());
+            allTasks.addAll(restoredActiveTasks);
+            return Collections.unmodifiableSet(allTasks);
+        } finally {
+            tasksAndActionsLock.unlock();
+            restoredActiveTasksLock.unlock();
+        }
+    }
+
+    @Override
+    public void shutdown(final Duration timeout) {
+        if (stateUpdaterThread != null) {
+            stateUpdaterThread.isRunning.set(false);
+            stateUpdaterThread.interrupt();
+            try {
+                stateUpdaterThread.join(timeout.toMillis());
```

Comment:
Generally we use a latch to check if the thread terminates successfully in time since the `join` function returns void and we cannot check if it did die within the timeout. If it did not terminate within the timeout, we could consider throw an exception for it.